samza-api/src/main/java/org/apache/samza/system/StreamSpec.java

```diff
@@ -0,0 +1,159 @@
+   *
+   * @param physicalName  The physical identifier for the stream. This is the identifier that will be used in remote
+   *                      systems to identify the stream. In Kafka this would be the topic name whereas in HDFS it
+   *                      might be a file URN.
+   *
+   * @param systemName    The System name on which this stream will exist. Corresponds to a named implementation of the
+ *                        Samza System abstraction. See {@link SystemFactory}
+   *
+   * @param properties    A set of properties for the stream. These may be System-specfic.
+   */
+  public StreamSpec(String id, String physicalName, String systemName, Properties properties) {
+    if (id == null) {
+      throw new NullPointerException("Parameter 'id' must not be null");
+    }
+
+    if (systemName == null) {
+      throw new NullPointerException("Parameter 'systemName' must not be null");
+    }
+
+    this.id = id;
+    this.systemName = systemName;
+    this.physicalName = physicalName;
+
+    if (properties != null) {
+      this.properties.putAll(properties);
+    }
+  }
+
+  /**
+   *  @param id             The application-unique logical identifier for the stream. It is used to distinguish between
+   *                        streams in a Samza application so it must be unique in the context of one deployable unit.
+   *                        It does not need to be globally unique or unique with respect to a host.
+   *
+   * @param physicalName    The physical identifier for the stream. This is the identifier that will be used in remote
+   *                        systems to identify the stream. In Kafka this would be the topic name whereas in HDFS it
+   *                        might be a file URN.
+   *
+   * @param systemName      The System name on which this stream will exist. Corresponds to a named implementation of the
+   *                        Samza System abstraction. See {@link SystemFactory}
+   *
+   * @param partitionCount  The number of partitionts for the stream. A value of {@code 1} indicates unpartitioned.
+   *
+   * @param properties      A set of properties for the stream. These may be System-specfic.
+   */
+  public StreamSpec(String id, String physicalName, String systemName, int partitionCount, Properties properties) {
+    this(id, physicalName, systemName, properties);
+    this.partitionCount = partitionCount;
+  }
+
+  public String getId() {
+    return id;
+  }
+
+
+  public String getSystemName() {
+    return systemName;
+  }
+
+  public String getPhysicalName() {
+    return physicalName;
+  }
+
+  public void setPhysicalName(String physicalName) {
+    this.physicalName = physicalName;
+  }
+
+  public int getPartitionCount() {
+    return partitionCount;
+  }
+
+  public void setPartitionCount(int partitionCount) {
+    this.partitionCount = partitionCount;
+  }
+
+  public Properties getProperties() {
+    return properties;
```

Comment:
I think the latest patch addresses everything. 

Note: in order to avoid confusion RE: Properties vs. Map, I renamed this field to "config". Not sure I like it, so I welcome suggestions.